controllers/s3/s3bucket.go

```diff
@@ -0,0 +1,236 @@
+// calls Observe in order to determine whether an external resource needs to be
+// created, updated, or deleted.
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) {
+	bucket, ok := mg.(*cloudscalev1alpha1.S3Bucket)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotInstance)
+	}
+	log.Info("Observe", "bucket", bucket.Name)
+
+	bucketUser, err := e.s3Client.GetBucketInfo(ctx, bucket.Status.ObjectUserID, bucket.GetBucketName())
+
+	// If we encounter an error indicating the external resource does not exist
+	// we want to let the resource.ManagedReconciler know so it can create it.
+	if s3.IsErrorNotFound(err) {
+		return resource.ExternalObservation{ResourceExists: false}, nil
+	} else if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(err, "cannot get instance")
+	}
+
+	// Update our "Ready" status condition to reflect the status of the external
+	// resource. Most managed resources use the below well known reasons that
+	// the "Ready" status may be true or false, but managed resource authors
+	// are welcome to define and use their own.
+	switch bucket.Status.Status {
+	case statusOnline:
+		// If the resource is available we also want to mark it as bindable to
+		// resource claims.
+		bucket.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(bucket)
+	case statusCreating:
+		bucket.SetConditions(runtimev1alpha1.Creating())
+	case statusDeleting:
+		bucket.SetConditions(runtimev1alpha1.Deleting())
+	default:
+		bucket.Status.Status = statusCreating
+	}
+
+	accessKey, secretKey, err := s3.GetKeys(bucketUser)
+	if err != nil {
+		return resource.ExternalObservation{}, err
+	}
+
+	// Finally, we report what we know about the external resource. Any
+	// ConnectionDetails we return will be published to the managed resource's
+	// connection secret if it specified one.
+	o := resource.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: true,
+		ConnectionDetails: resource.ConnectionDetails{
+			runtimev1alpha1.ResourceCredentialsSecretUserKey:     []byte(accessKey),
+			runtimev1alpha1.ResourceCredentialsSecretPasswordKey: []byte(secretKey),
+			runtimev1alpha1.ResourceCredentialsSecretEndpointKey: []byte(cloudscale.S3Endpoint),
+		},
+	}
+
+	return o, nil
+}
+
+// Create a new external resource based on the specification of our managed
+// resource. resource.ManagedReconciler only calls Create if Observe reported
+// that the external resource did not exist.
+func (e *external) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	bucket, ok := mg.(*cloudscalev1alpha1.S3Bucket)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errNotInstance)
+	}
+	log.Info("Create", "bucket", bucket.Name)
+
+	objectUser, err := e.s3Client.CreateOrUpdateBucket(ctx, bucket.Status.ObjectUserID, bucket.GetBucketName(), bucket.Spec.Tags)
```

Comment:
We have begun to standardize around using an annotation to represent the external name of a resource. Because it appears that this provider allows you to specify external name before creation, the annotation should be set on the managed resource using `meta.SetExternalName()` and retrieved via `meta.GetExternalName()`.

Doc Ref: https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#external-resource-name

Reply:
Not sure I got this right. Like that?

Reply:
By default, the following initializer is assigned to managed reconciler https://github.com/crossplaneio/crossplane-runtime/blob/master/pkg/resource/managed_reconciler.go#L259

It basically uses `<claim namespace>-<claim-name>-<5 rand characters>` format to create a name and assigns it as an annotation on the managed resource. What you need to do is simply calling `meta.GetExternalName(cr)` like here https://github.com/crossplaneio/stack-aws/blob/master/pkg/controller/cache/managed.go#L182 whenever you need the name of the resource.

Reply:
But what if I want to provide the ability to set the name of the external resource in a claim?

Reply:
If your claim has an annotation with key `crossplane.io/external-name`, its value will be used instead of generating a new one.

Nevertheless, you can always use any mechanism you want. This external name propagation mechanism is used across managed resources but if you'd like a field to represent it, that'd also work.

Reply:
Note: If you do implement your own mechanism you would want to disable the annotation one:

`resource.WithManagedInitializers(resource.NewAPIManagedFinalizerAdder(mgr.GetClient()))`

I recently did this here: https://github.com/packethost/stack-packet/pull/9/files#diff-57be8f6c5cd137378ee9756c8ef40727R66

So that I could set the external name after a device was created in `stack-packet`: https://github.com/packethost/stack-packet/pull/9/files#diff-57be8f6c5cd137378ee9756c8ef40727R178

Reply:
> Note: If you do implement your own mechanism you would want to disable the annotation one.

I'm not sure that's true in this case. You need to disable the annotation mechanism for managed resources with non-deterministic external names generated by the external system, to ensure we propagate the externally generated external name back to the annotation.

As @muvaf pointed out, in the default case where you do not disable the external naming functionality by overriding the initializers:

1. Dynamically provisioned managed resources (i.e. those provisioned by a claim) will have their `crossplane.io/external-name` annotation set to the claim's `crossplane.io/external-name` annotation, if any. This happens in the [`ObjectMetaConfigurator`](https://godoc.org/github.com/crossplaneio/crossplane-runtime/pkg/resource#ObjectMetaConfigurator).
1. Statically provisioned managed resources, and dynamically provisioned resources whose claim did not specify an `crossplane.io/external-name` annotation will set that annotation to the name of the managed resource. This happens in the [`ManagedNameAsExternalName`](https://godoc.org/github.com/crossplaneio/crossplane-runtime/pkg/resource#ManagedNameAsExternalName) initializer.

So currently I believe this implementation will:

1. Use the claim's `crossplane.io/external-name` annotation as the bucket name if it is set.
1. Fall back to using a name determined by the `NameFormat` as the bucket name.

I was hoping to deprecate the `NameFormat` pattern - it's an older pattern we used to generate random names for dynamically provisioned managed resources before `ExternalName` gave more control over naming. If you removed the `NameFormat` support and instead used the default `meta.GetExternalName` behaviour the implementation would:

1. Use the claim's `crossplane.io/external-name` annotation as the bucket name if it is set.
1. Fall back to using the managed resource's name, which is `claimnamespace-claimname-randomstring` for dynamically provisioned resources, or whatever name you choose for statically provisioned resources.

Do you have a preference for the older `NameFormat` style?

Reply:
Ok, now I got it. I'll change to your proposed behavior regarding `external-name` and remove `NameFormat` style :+1: